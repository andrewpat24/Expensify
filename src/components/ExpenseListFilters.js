import React from 'react';
import { connect } from 'react-redux';
// Helpers
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters';

const ExpenseListFilters = (props) => (
    <div>
        <input 
        type="text" 
        value={props.filters.text} 
        onChange={(e) => {
            props.dispatch(setTextFilter(e.target.value))
            console.log(e.target.value);
        }}/>
        <select 
        value={props.filters.sortBy}
        onChange={(e) => {
            console.log('select changed.', e.target.value);
            e.target.value==="Date" ? props.dispatch(sortByDate()) : props.dispatch(sortByAmount())
        }}>
            <option>Date</option>
            <option>Amount</option>
        </select>
    </div>
)

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilters); 