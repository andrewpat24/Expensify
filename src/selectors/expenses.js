const getVisibleExpenses = (
  expenses,
  /* filters:*/ { text, sortBy, startDate, endDate }
) => {
  console.log(
    // `Expenses: ${expenses} \n`,
    `text: ${text}, sortby: ${sortBy}, startDate: ${startDate}, endDate: ${endDate}`
  );
  return expenses
    .filter(expense => {
      const startDateMatch = !startDate || expense.createdAt >= startDate;
      const endDateMatch = !endDate || expense.createdAt <= endDate;
      const textMatch = expense.description.toLowerCase().includes(text);
      console.log(expense, startDateMatch && endDateMatch && textMatch);
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
