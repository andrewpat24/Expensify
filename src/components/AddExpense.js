import React from 'react';
import { connect } from 'react-redux';
// Components
import ExpenseForm from './ExpenseForm';
// Helpers
import { addExpense } from '../actions/expenses';

const AddExpense = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm 
        onSubmit={(expense) => {
            props.dispatch( addExpense(expense) );
            props.history.push('/');
        }}
        /> 
    </div>
)


export default connect()(AddExpense);