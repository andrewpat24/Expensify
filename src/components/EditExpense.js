import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
// Helpers
import { editExpense, removeExpense } from "../actions/expenses";

const EditExpense = props => {
  if (typeof props.expense === "undefined") {
    return props.history.push("/");
  }

  return (
    <div>
      <h1>Editing doc id: {props.match.params.id}</h1>
      <h2>{props.expense.description}</h2>
      <ExpenseForm
        expense={props.expense}
        expenseIsNew={false}
        onSubmit={expense => {
          props.dispatch(editExpense(props.expense.id, expense));
          props.history.push("/");
        }}
      />
      <button
        onClick={() => {
          props.dispatch(removeExpense({ id: props.expense.id }));
          props.history.push("/");
        }}
      >
        Remove
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

export default connect(mapStateToProps)(EditExpense);
