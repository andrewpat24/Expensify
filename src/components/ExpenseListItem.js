import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
// Helpers
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = (props) => {
    const { dispatch, id, description, amount=100, createdAt=500 } = props;  
    return (
            <div>
                <Link to={`/edit/${id}`}>
                    <h1>{description}</h1>
                </Link>
                <h2>Amount: {amount}</h2>
                <h2>Created at: {createdAt}</h2>
                
                <button onClick={ () => { dispatch( 
                    removeExpense({
                        id
                    }) 
                ) } }>Remove</button>
            </div>
    )
}

export default connect()(ExpenseListItem); 