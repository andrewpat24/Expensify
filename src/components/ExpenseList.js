import React, { Fragment } from "react";
import { connect } from "react-redux";
// Components
import ExpenseListItem from "./ExpenseListItem";
// Helpers
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      <div>
        {props.expenses.map(expense => {
          return <ExpenseListItem {...expense} key={expense.id} />;
        })}
      </div>
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
