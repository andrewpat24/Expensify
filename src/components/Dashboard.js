import React from 'react';
// Components 
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const Dashboard = () => (
    <section component="Dashboard">
        <h1>Dashboard</h1>
        <ExpenseListFilters /> 
        <ExpenseList />
    </section>
)

export default Dashboard; 