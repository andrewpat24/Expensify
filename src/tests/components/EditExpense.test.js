import React from "react";
import { shallow } from "enzyme";
import { EditExpense } from "../../components/EditExpense";
import expenses from "../fixtures/expenses";

let editExpense, removeExpense, history, wrapper;
const { amount, createdAt, description, id, note } = expenses[0];

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpense
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test("Should render EditExpense component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenCalledWith("/");
  expect(editExpense).toHaveBeenLastCalledWith(id, {
    amount,
    createdAt,
    description,
    id,
    note
  });
});

test("Should handle removeExpense", () => {
  wrapper.find("button").prop("onClick")(expenses[0]);
  expect(history.push).toHaveBeenCalledWith("/");
  expect(removeExpense).toHaveBeenLastCalledWith({
    id
  });
});
