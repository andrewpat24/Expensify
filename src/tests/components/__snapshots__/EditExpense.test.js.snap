// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpense component correctly 1`] = `
<div>
  <h1>
    Editing doc id: 
    1
  </h1>
  <h2>
    Gum
  </h2>
  <ExpenseForm
    expense={
      Object {
        "amount": 50,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "Test note",
      }
    }
    expenseIsNew={false}
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
